ann_benchmarks/algorithms/dragonfly_dev/module.py

```diff
@@ -113,60 +115,25 @@ def query(self, v, n):
         ]
         resp = self.redis.execute_command(*q, target_nodes="random")
 
-        # Robustly extract document IDs from RediSearch response
-        ids = []
-        if isinstance(resp, (list, tuple)) and len(resp) > 1:
-            for item in resp[1:]:
-                if isinstance(item, (bytes, str)):
-                    try:
-                        ids.append(int(item))
-                    except Exception:
-                        continue
-                elif isinstance(item, (list, tuple)) and len(item) > 0:
-                    doc_id = item[0]
-                    if isinstance(doc_id, (bytes, str)):
-                        try:
-                            ids.append(int(doc_id))
-                        except Exception:
-                            continue
-        return ids
+        # Extract document IDs from RediSearch response
+        return [int(resp[i]) for i in range(1, len(resp), 2)] if len(resp) > 1 else []
 
     def batch_query(self, X, n):
-        """Run many queries concurrently using a thread pool.
+        """Run many queries using Redis pipelining for optimal performance.
 
-        Each thread holds its own Redis client bound to a single dedicated TCP connection.
-        This enables true I/O parallelism across many sockets.
+        Uses Dragonfly's native batch processing capabilities via Redis pipeline
+        instead of client-side threading to reduce network overhead and improve throughput.
         """
         total = len(X)
         self._batch_results = [None] * total
         self._batch_latencies = [0.0] * total
 
-        tls = threading.local()
-
-        def get_client() -> Redis:
-            # Lazily create one Redis client per OS thread.
-            cli = getattr(tls, "cli", None)
-            if cli is None:
-                cli = Redis(
-                    host=self.host,
-                    port=self.port,
-                    decode_responses=False,
-                    socket_timeout=self.search_read_timeout,
-                    socket_connect_timeout=self.connect_timeout,
-                    socket_keepalive=True,
-                    health_check_interval=30,
-                    retry_on_timeout=True,
-                )
-                # warm up connection
-                try:
-                    cli.ping()
-                except Exception:
-                    pass
-                tls.cli = cli
-            return cli
-
-        def run_one(idx: int):
-            v = X[idx].astype(numpy.float32)
+        # Create pipeline for batch execution
+        pipeline = self.redis.pipeline(transaction=False)
```

Comment:
I am sorry, but I do not think @dranikpg  suggestion was good. Pipeline use a single connection so, with a single global index you won't see any difference vs using sharded indices. I still think we should use multiple connections in order to load Dragonfly (either main or the POC).